feat(aria): add aria labeling to the trigger

jason-capsule42 opened the pull request:
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Fixes:** #77

## Summary:

This change is meant to get a valid aria labeling rule on the trigger element.

If a custom `aria-label` is applied at the root that will get applied to the trigger elem.
Else if a custom `aria-labeledby` is applied to the root that will get applied.
Else if the `label` slot has content that will get applied.

After reviewing the implementation of this with @geoffrich it was understood that *all* instances of `auro-dropdown` need to have `aria-label` or `aria-labelledby` on the trigger or the voiceover experience will not meet expectations. An update has been made to the demo page explaining this and all apiExamples have been updated to be compliant with this expectations.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [ ] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)


## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -94,7 +94,32 @@ class AuroDropdown extends LitElement {
+    if (this.getAttribute('aria-label')) {
+      // If custom aria-label declared at root apply it to the trigger
+      this.trigger.setAttribute('aria-label', this.getAttribute('aria-label'));
+    } else if (this.getAttribute('aria-labeledby')) {
```
This attribute is misspelled - it's [aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby)

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -94,7 +94,32 @@ class AuroDropdown extends LitElement {
+
+    if (this.getAttribute('aria-label')) {
+      // If custom aria-label declared at root apply it to the trigger
+      this.trigger.setAttribute('aria-label', this.getAttribute('aria-label'));
```
I don't like imperatively setting the attribute like this. Can we do it in the template instead?

jason-capsule42 replied:
Can you expand on a reason to move this more than "don't like"?

We can certainly move it to the template but I think that code is likely heavier and requires a little more mental leaps to account for the logic of when to use `label` vs `labelledby` and not end up with both attributes on the trigger.

geoffrich replied:
Sorry, I can clarify - I'm fine with keeping the if/else logic here. But I don't think we should be imperatively setting attributes inside the component with `this.trigger.setAttribute`. In general, if we're modifying the component's template, we should do it in the render method. Doing it imperatively in a different method can make it difficult to track down where it's happening. Instead, we should set a state variable and then apply it in the render method. So I'd prefer something like this... 

```
else if (this.getAttribute('aria-labelledby')) {
      // Else if custom aria-labelledby declared at root apply it to trigger
      this.labelledBy = this.getAttribute('aria-labelledby');
    }
// ...then in the render method

<div id="trigger" class="trigger" data-trigger-placement="${this.placement}" aria-labelledby=${this.labelledBy}>
```

I think doing it this way is also more performant, since Lit can optimize the state updates, though I'm not 100% certain.

Initializing `this.labelledBy` to undefined should make sure the attribute is not set by default (IIRC).

jason-capsule42 replied:
@geoffrich This change was achieved using `ifDefined` rules in the template along with the change to reference properties rather than apply the attributes directly with JavaScript.

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -94,7 +94,32 @@ class AuroDropdown extends LitElement {
+      this.trigger.setAttribute('aria-label', this.getAttribute('aria-label'));
+    } else if (this.getAttribute('aria-labeledby')) {
+      // Else if custom aria-labeledby declared at root apply it to trigger
+      this.trigger.setAttribute('aria-labeledby', this.getAttribute('aria-labeledby'));
```
I'm not sure if this will be valid. We should test this behavior with a screen reader. My concern is that `<auro-dropdown aria-labelledby="foo">` will refer to an element with ID "foo" outside of this component's shadow root, and using `aria-labelledby="foo"` inside the shadow root won't be able to reference that element.

jason-capsule42 replied:
This works correctly for me using VoiceOver on my mac. If you run the localhost demo page that last example in the demo renders with this method. @geoffrich please let me know if your mileage is different with this in the windows environment! I'd appreciate the feedback after trying out that coded example.

geoffrich replied:
This only worked once I added `role="button"` to the trigger. Otherwise, NVDA wouldn't announce the labels (presumably because the trigger wasn't exposed as an interactive element). Can you make that change?

jason-capsule42 commented:
@blackfalcon I am done making updates to this PR based on the conversation with @geoffrich. I think this is ready for a final review from you.

geoffrich commented on demo/demo.md:
```diff
@@ -32,10 +32,20 @@ The dropdown can be closed with the following actions:
 
+## Accessibility requirements
+
+All instances of `auro-dropdown` must have a valid `aria-label` or `aria-labelledby` attribute. There are three options available to meet this requirement. Only one of these options should be implemented.
```
This is incorrect - the label slot is also an option, and doesn't require the dev to set aria-label or aria-labelledby themselves. Aria attributes should only be used when absolutely necessary. Can we re-word this to suggest the label slot as the primary method of labelling the dropdown? 

I'd also recommend updating the examples to use the label slot instead of aria-label.

jason-capsule42 replied:
@geoffrich The way I wrote this is meant to say that one way or another you must get an `aria-label` or `aria-labelledby` attribute on the trigger. Using the slot is just another means of making that end goal happen.

Regarding if the slot should be our default example or not - I will bring this up during our standup discussion today with @leeejune. I'd like design to weigh in here. I had understood the label slot to be optional but not the default implementation choice.

geoffrich replied:
Ah, gotcha - we can go with whatever design says on that front. I still think this is worded confusingly; the user of the component shouldn't care that the label slot sets aria-labelledby behind the scenes.

jason-capsule42 replied:
I agree the wording could improve. I put a tweak in that may be less confusing by saying less. :)

Let me know what you think, @geoffrich.

geoffrich replied:
Better, but I think we can make it even simpler. How about "All instances of `auro-dropdown` must have a valid label." for the first sentence? Then the rest is good

geoffrich commented on demo/demo.md:
```diff
@@ -32,10 +32,20 @@ The dropdown can be closed with the following actions:
+- Add the `aria-labelledby` attribute to the dropdown root (e.g. `<auro-dropdown aria-labelledby="idOfElement">`)
+- Include content in the `label` slot. When doing so, the `aria-labelledby` attribute will automatically reference the label slot content.
+
+Failing to include one of the above options will result in voiceover utilities being unable to correctly identify the dropdown to the user.
```
```suggestion
Failing to include one of the above options will prevent assistive technology from correctly identifying the dropdown to the user.

```

jason-capsule42 replied:
I like the rewording here, @geoffrich. I was struggling trying to find the right words. Great suggestion. Thank you.

geoffrich commented on docs/partials/demo.md:
```diff
@@ -27,6 +27,16 @@ The dropdown can be closed with the following actions:
 
+## Accessibility requirements
+
+All instances of `auro-dropdown` must have a valid `aria-label` or `aria-labelledby` attribute. There are three options available to meet this requirement. Only one of these options should be implemented.
```
This is a duplicate of the content in `demo.md` - we should only have one.

jason-capsule42 replied:
This is the template that our compiled ./demo/demo.md builds from. This is where the change is made manually.

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -47,6 +48,7 @@ class AuroDropdown extends LitElement {
     this.inset = false;
     this.rounded = false;
     this.toggle = false;
+    // this.ariaLabel = 'test label from prop';
```
Remove commented out code

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -94,7 +106,32 @@ class AuroDropdown extends LitElement {
+   * @returns {void} Updates the trigger `aria-label` or `aria-labelledby`.
+   */
+  processAriaLabel() {
+    this.trigger = this.shadowRoot.querySelector(`#trigger`);
```
This is a duplicate of line 100 - can we remove this line and this function call after the trigger is set?

jason-capsule42 replied:
I may be misunderstanding here due to the line numbers changing after making updates from other commits, but I don't see where this is already declared elsewhere in the method.

geoffrich replied:
We also set the trigger in the `firstUpdated` method, where this method is called from. We should only set it once

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -219,13 +256,19 @@ class AuroDropdown extends LitElement {
+        class="trigger"
+        role="button"
+        aria-label="${ifDefined(this.ariaLabel ? this.ariaLabel : undefined)}"
+        aria-labelledby="${ifDefined(this.ariaLabelledby ? this.ariaLabelledby : undefined)}"
```
```suggestion
        aria-label="${ifDefined(this.ariaLabel)}"
        aria-labelledby="${ifDefined(this.ariaLabelledby)}"
```

I _think_ this should work instead - asumming ariaLabel and ariaLabelledby defaults to undefined

jason-capsule42 replied:
You were right, the shorthand version does work. Thanks, @geoffrich.

geoffrich commented on src/auro-dropdown.js:
```diff
@@ -219,13 +256,19 @@ class AuroDropdown extends LitElement {
           </div>
           <div class="triggerContent" chevron=${this.chevron}>
-            <slot name="trigger"></slot>
+            <slot name="trigger"></slot>${this.ariaLabel}
```
Why was this added?

blackfalcon reviewed (changes requested):
This branch and PR should be updated to be a PATCH release. The issue was reported as a bug and this update resolves an issue related to a broken experience, not add a new feature to the element.

blackfalcon committed:
fix(aria): add aria labeling to the trigger #77

Changes to be committed:
modified:   src/auro-dropdown.js

blackfalcon committed:
docs(*): update per new a11y spec update

Changes to be committed:
modified:   README.md
modified:   demo/apiExamples.md
modified:   demo/demo.md
modified:   docs/partials/demo.md

blackfalcon commented:
@geoffrich @jason-capsule42 

I am curious as to this implementation. When I made this issue and linked to the `aria-labeledby` recommendation, the intent that I had was that we would use this to bind the content from the string in the `label` slot to this aria rule. 

The error that I encountered was that since focusing on the trigger totally bypasses the label, it is never read back to the user. 

The following is the update I had in mind. 

```js
  render() {
    return html`
      <div id="popover" class="popover" aria-live="polite" style=${`min-width: ${this.dropdownWidth}px;`}>
        <slot role="tooltip"></slot>
      </div>
      <div
        id="trigger"
        class="trigger"
        role="button"
        data-trigger-placement="${this.placement}">
        <div class="triggerContentWrapper">
          <label class="label" id="triggerLabel">
            <slot name="label"></slot>
          </label>
          <div class="triggerContent" chevron=${this.chevron} aria-labeledby="triggerLabel">
            <slot name="trigger"></slot>
          </div>
        </div>
        ${this.chevron ? html`
          <div id="showStateIcon">
            <auro-icon
              category="interface"
              name="chevron-down"
              customColor>
            </auro-icon>
          </div>
        ` : undefined}
      </div>
      <div class="helpText">
        <slot name="helpText"></slot>
      </div>
    `;
  }
}
```

With this update, the content in the `label` slot will take priority over the use of `aria-label` at the root. But, using `aria-label` at the root address the use case for when the `label` slot is not used. 

Where this does fall short is using `aria-labeledby` to content outside the scope of the element. But I am not sure that is a use case we need to support. Either the `label` slot is used or `aria-label` on the root is used instead of label content. 

What is the advantage of the current solution versus what I am proposing?

blackfalcon committed:
refactor(feature): reduce scope of code to address solution

Changes to be committed:
modified:   apiExamples/commonAdvanced.html
modified:   apiExamples/commonBasic.html
modified:   src/auro-dropdown.js
modified:   test/auro-dropdown.test.js

jason-capsule42 commented on package.json:
```diff
@@ -142,6 +142,7 @@
     "build:dev:assets": "npm-run-all build:sass:demo build:sass:component postCss:component sass:render",
     "build:markdownDocs": "node scripts/generateDocs.js",
     "build:sass": "npm-run-all build:sass:demo build:sass:component postCss:component sass:render",
+    "build:md": "npm run build:markdownDocs",
```
This change should not be part of this PR. If we want to make a shorthand of `build:markdownDocs` (which I am in favor of) that change needs to be in WC-Generator.

jason-capsule42 committed:
docs(examples): fix broken API example code

jason-capsule42 committed:
test(aria): restore aria-labelledby test

jason-capsule42 commented:
@blackfalcon The update you made has two notable impacts:

1. The `aria-labelledby` attribute is left on the trigger content even when it's not going to be in use. This doesn't seem to cause a problem for VoiceOver on my mac. @geoffrich would you be able to check using NVDA?
1. It's no longer possible for a custom `aria-labelledby`. I'm not sure why we would choose to support a custom `aria-label` but not `aria-labelledby`.

Otherwise, the change seems fine.

blackfalcon commented:
What is the scenario for a custom labeledby? 

The only thing I can think of is attaching this to something else in the DOM outside the scope of the element. If that is the case, that has to happen on the parent container of the element, aria does not pierce the shadow DOM.

blackfalcon committed:
build: update scripts to automate generation of md docs on change

Changes to be committed:
modified:   package.json

blackfalcon committed:
docs(demo): update to support different label scenarios

Changes to be committed:
renamed:    apiExamples/commonBasic.html -> apiExamples/common.html
new file:   apiExamples/commonLabelledby.html
renamed:    apiExamples/commonAdvanced.html -> apiExamples/commonSlot.html
modified:   demo/demo.md
modified:   docs/partials/demo.md

jason-capsule42 commented on demo/demo.md:
```diff
@@ -89,17 +127,22 @@ Failing to include one of the above options will prevent assistive technology fr
 
 </auro-accordion>
+
+## Common use using the label content slot
```
"Common use with the label content slot" is consistent with the headers on other examples
```suggestion
## Common use with the label content slot
```

jason-capsule42 commented on demo/demo.md:
```diff
@@ -138,3 +181,18 @@ Failing to include one of the above options will prevent assistive technology fr
+
+The dropdown can be closed with the following actions:
+
+1. Clicking anywhere in the view outside of the dropdown.
```
How about `document` or `window` instead of `view` which I don't think is a correct term here?

jason-capsule42 commented on demo/demo.md:
```diff
@@ -138,3 +181,18 @@ Failing to include one of the above options will prevent assistive technology fr
+
+1. Clicking anywhere in the view outside of the dropdown.
+1. When the `toggle` attribute is added, clicking on the trigger.
+1. Programmatically via another control in the UI calling the `hide()` method (see api).
```
Doesn't have to be from a UI control.
```suggestion
1. Programmatically via calling the `hide()` method (see api).
```

jason-capsule42 commented on demo/demo.md:
```diff
@@ -138,3 +181,18 @@ Failing to include one of the above options will prevent assistive technology fr
+1. Clicking anywhere in the view outside of the dropdown.
+1. When the `toggle` attribute is added, clicking on the trigger.
+1. Programmatically via another control in the UI calling the `hide()` method (see api).
+1. Selecting an option within the dropdown by either keyboard or click/tap.
```
There is no code in dropdown itself that does this. For instance in the case of `auro-select` making a selection from the menu calls `dropdown.hide()`.

We should remove this line.

blackfalcon commented on demo/demo.md:
```diff
@@ -128,3 +181,18 @@ The dropdown can be closed with the following actions:
+
+The dropdown can be closed with the following actions:
+
+1. Clicking anywhere in the view outside of the dropdown.
```
```suggestion
1. Clicking anywhere in the wndow outside of the dropdown.
```

blackfalcon commented on demo/demo.md:
```diff
@@ -128,3 +181,18 @@ The dropdown can be closed with the following actions:
+1. Clicking anywhere in the view outside of the dropdown.
+1. When the `toggle` attribute is added, clicking on the trigger.
+1. Programmatically via another control in the UI calling the `hide()` method (see api).
+1. Selecting an option within the dropdown by either keyboard or click/tap.
```
```suggestion
```

blackfalcon committed:
refactor(*): apply updates per review

Co-authored-by: Jason Baker <jason@capsule42.com>

blackfalcon commented:
:tada: This PR is included in version 2.0.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@aurodesignsystem/auro-dropdown/v/2.0.1)
- [GitHub release](https://github.com/AlaskaAirlines/auro-dropdown/releases/tag/v2.0.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
